packages/web5/lib/src/pexv2/pd.dart

```diff
@@ -0,0 +1,274 @@
+
+/// PresentationDefinition represents a DIF Presentation Definition defined [here].
+/// Presentation Definitions are objects that articulate what proofs a Verifier requires.
+///
+/// [here]: https://identity.foundation/presentation-exchange/#presentation-definition
+class PresentationDefinition {
+  String id;
+  String? name;
+  String? purpose;
+  List<InputDescriptor> inputDescriptors;
+
+  PresentationDefinition({
+    required this.id,
+    this.name,
+    this.purpose,
+    required this.inputDescriptors,
+  });
+
+  factory PresentationDefinition.fromJson(Map<String, dynamic> json) =>
+      PresentationDefinition(
+        id: json['id'],
+        name: json['name'],
+        purpose: json['purpose'],
+        inputDescriptors: List<InputDescriptor>.from(
+          json['input_descriptors'].map((x) => InputDescriptor.fromJson(x)),
+        ),
+      );
+
+  Map<String, dynamic> toJson() => {
+        'id': id,
+        'name': name,
+        'purpose': purpose,
+        'input_descriptors':
+            List<dynamic>.from(inputDescriptors.map((x) => x.toJson())),
+      };
+
+  List<String> selectCredentials(List<String> vcJwts) {
+    final Set<String> matches = {};
+
+    for (final inputDescriptor in inputDescriptors) {
+      final matchingVcJwts = inputDescriptor.selectCredentials(vcJwts);
+      if (matchingVcJwts.isEmpty) {
+        return [];
+      }
+      matches.addAll(matchingVcJwts);
+    }
+
+    return matches.toList();
+  }
+}
+
+class _TokenizedField {
+  List<String> paths;
+  String token;
+
+  _TokenizedField({required this.paths, required this.token});
+}
+
+/// InputDescriptor represents a DIF Input Descriptor defined [here].
+/// Input Descriptors are used to describe the information a Verifier requires of a Holder.
+///
+/// [here]: https://identity.foundation/presentation-exchange/#input-descriptor
+class InputDescriptor {
+  String id;
+  String? name;
+  String? purpose;
+  Constraints constraints;
+
+  InputDescriptor({
+    required this.id,
+    this.name,
+    this.purpose,
+    required this.constraints,
+  });
+
+  factory InputDescriptor.fromJson(Map<String, dynamic> json) =>
+      InputDescriptor(
+        id: json['id'],
+        name: json['name'],
+        purpose: json['purpose'],
+        constraints: Constraints.fromJson(json['constraints']),
+      );
+
+  Map<String, dynamic> toJson() => {
+        'id': id,
+        'name': name,
+        'purpose': purpose,
+        'constraints': constraints.toJson(),
+      };
+
+  String _generateRandomToken() {
+    final rand = Random.secure();
+    final bytes = Uint8List(16);
+    for (int i = 0; i < 16; i++) {
+      bytes[i] = rand.nextInt(256);
+    }
+    return hex.encode(bytes);
+  }
+
+  List<String> selectCredentials(List<String> vcJWTs) {
+    final List<String> answer = [];
+    final List<_TokenizedField> tokenizedField = [];
+    final schemaMap = {
+      '\$schema': 'http://json-schema.org/draft-07/schema#',
```

Comment:
why slash?

also you should not need to construct a schema like this - the schema should be conformant in the PD itself

Reply:
`$` is used for string interpolation in dart, so `\$` escapes the `$` so we can put an actual `$` in the string.

What schema in the PD are you referring to?

Reply:
I am working off the structure of these test vectors from the web5-spec. Are these incorrect? https://github.com/TBD54566975/web5-spec/blob/main/test-vectors/presentation_exchange/select_credentials.json#L11-L26

Reply:
those look correct, though I do not think you need the `schemaMap` to build a schema ... the value of a `filter` itself should be a valid json schema as per https://identity.foundation/presentation-exchange/#input-descriptor-object

Reply:
@decentralgabe take a look at the algorithm discussed here: https://github.com/TBD54566975/web5-dart/issues/52

it significantly cut down on the amount of nested loops we've seen in other SDKs and simplified PEX eval overall

Reply:
also @diehuxx given that you already have a strongly typed `Filter`, you could easily create a minimal strongly typed `JsonSchema` class

Reply:
Added! I opted to make the strongly typed `_JsonSchema` class private since its functionality is pretty specific to what we need for this `selectCredentials` algo.